### PR TITLE
[Feature] Add preserveModulesRoot config option

### PR DESCRIPTION
**Pull request by davidroeca**
This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### Description

This PR adds a feature mentioned in #3785 that allows users to specify a `moduleRootDir` option. This allows users to specify how specific input modules' paths should be rendered in the output module. It is merely one implementation of multiple alternatives, so I figured I would draft up this PR as one resolution to #3785, but there may be better solutions for this.

### CODE REVIEW

**Comment by lukastaegert** [comment]
Thanks for the contribution, that is actually an interesting proposal. So if I understand this correctly what it does is:
If
- we are preserving modules and
- an id starts with the "moduleRootDir" option value,

then
- replace it with an empty string

So my initial thoughts were:
- that is actually a very simple and non-breaking approach, nicely done!
- Is it configurable enough, i.e. are there situations where I want more control? But on second thought I think it is, at least I could not think of scenarios where I want more control. And there is already a note to give full control over the path names when preserving modules in Rollup 3.

So if you would be willing to flesh out this PR, I would be willing to release it.

**Review by lukastaegert — CHANGES_REQUESTED** [review]
Besides the comments that I left, these are the things that need to be done:
- Document the option in `docs/999-big-list-of-options.md`. I think it would be an "advanced" option. Mind the alphabetic ordering! It would also need to be added to the "Configuration Files" and "Command Line Flags" sections of `docs/01-command-line-reference.md`, the "inputOptions object" section of `docs/02-javascript-api.md` and the `cli/help.md` file
- Fix the existing tests. In this case, all you need to do is add the new option in the `test/misc/optionList.js` in the appropriate places.
- Add a test. This one would go into the `chunking-form` category. If we allow both strings and regular expressions, then these would be two tests, one for each, and hopefully the regular expression test does something interesting that is not possible with strings...

File: src/Chunk.ts [hunk 1]
```diff
@@ -441,20 +441,30 @@ export default class Chunk {
...
+					ext: () => extension.substr(1),
+					extname: () => extension,
+					format: () => options.format as string,
+					name: () => this.getChunkName()
+				},
+				this.getChunkInfo.bind(this)
 			);
+			if (options.moduleRootDir) {
+				const moduleRootDir = resolve(options.moduleRootDir);
+				const moduleRootDirRegExp = new RegExp(`^${moduleRootDir}`);
```
[the remark this answers is not in the document]
Comment (davidroeca) [reply src/Chunk.ts:458]: Thanks for the review! Will start looking into these changes. With this one, I wasn't planning on using regex as an input option, so I think `.startsWith` is definitely preferable

File: src/rollup/types.d.ts [hunk 2]
```diff
@@ -575,6 +575,7 @@ export interface OutputOptions {
 	intro?: string | (() => string | Promise<string>);
 	manualChunks?: ManualChunksOption;
 	minifyInternalExports?: boolean;
+	moduleRootDir?: string;
```
[the remark this answers is not in the document]
Comment (davidroeca) [reply src/rollup/types.d.ts:578]: :+1: I like that name better

File: src/Chunk.ts [hunk 3]
```diff
@@ -441,20 +441,29 @@ export default class Chunk {
...
+				{
+					ext: () => extension.substr(1),
+					extname: () => extension,
+					format: () => options.format as string,
+					name: () => this.getChunkName()
+				},
+				this.getChunkInfo.bind(this)
 			);
+			if (options.preserveModulesRoot) {
+				const preserveModulesRoot = resolve(options.preserveModulesRoot);
```
Comment (lukastaegert) [inline src/Chunk.ts:457]: There is quite a bit of `resolving` going on here which is not free as `resolve` involves non-trivial file-system operations. So one way to get rid of the first one at least would be to perform this resolution once in `normalizeOutputOptions` so that `options.preserveModulesRoot` is already resolved at this point.

File: src/Chunk.ts [hunk 4]
```diff
@@ -441,20 +441,29 @@ export default class Chunk {
...
+				if (currentDir.startsWith(preserveModulesRoot)) {
+					currentDir = resolve(
+						preserveModulesRelativeDir,
+						currentDir.slice(preserveModulesRoot.length).replace(/^\//, '')
+					);
+				}
+			}
+			const currentPath = `${currentDir}/${fileName}`;
+			path = relative(preserveModulesRelativeDir, currentPath);
```
Comment (lukastaegert) [inline src/Chunk.ts:466]: I am thinking, is this really all necessary? Wouldn't this be the same:

```js
if (currentDir.startsWith(preserveModulesRoot)) {
  path = currentDir.slice(preserveModulesRoot.length).replace(/^\//, '')
} else {
  path = relative(preserveModulesRelativeDir, `${currentDir}/${fileName}`);
}
```

or am I overlooking something here?

Comment (davidroeca) [reply src/Chunk.ts:466]: In practice, this will probably will be the case, at least for right now. We would need to add a check somewhere to make sure `preserveModulesRelativeDir` is resolved as a parent directory to `preserveModulesRoot` for this to work out.

**Review by lukastaegert — APPROVED** [review]
Thanks a lot! I also very much like the test as you made sure that the ciritical edge cases of both node_modules and virtual modules are covered as well.

**Comment by dasa** [comment]
> * Document the option in `docs/999-big-list-of-options.md`.

Looks like this is missing.

**Comment by lukastaegert** [comment]
Ups, thanks for noticing. I guess we can add this is an additional docs PR. As always if someone could put a few lines together in a separate docs PR, that would be awesome. Otherwise I can take a look tomorrow.
